articles/user-guide/testing-debugging.md

```diff
@@ -349,20 +421,47 @@ The following examples show `DumpMachine` for some common states:
+
+In this case, for <xref:Microsoft.Quantum.Diagnostics.DumpMachine>, we explicitly show the development on Q# Jupyter Notebook as it offers more visualization tools for testing and debugging quantum programs.
 
-    operation Operation () : Unit {
+1. To run `DumpMachine` on Jupyter Notebook, open a [new Jupyter Notebook with a Q# kernel](xref:microsoft.quantum.install-qdk.overview.jupyter) and copy the following code to the first notebook cell.
+
+ ```qsharp
+open Microsoft.Quantum.Diagnostics;
+ 
+operation MultiQubitDumpMachineDemo() : Unit {
         use qubits = Qubit[2];
+        X(qubits[1]);
         H(qubits[1]);
-        DumpMachine("dump.txt");
+        R1Frac(1, 2, qubits[1]);
+        
+        DumpMachine();
+
+        ResetAll(qubits);
     }
-}
+
 ```
+1. In a new cell, run `MultiQubitDumpMachineDemo` operation on a full state quantum simulator by using the `%simulate` magic command. `DumpMachine` call will print the information about the quantum state of the program after the Controlled Ry gate as a set of lines, one per basis state, showing their complex amplitudes, phases, and measurement probabilities:
+
+![DumpMachine output](~/media/dumpmachine-output.png)
+
+> [!NOTE]
+> You can use <xref:microsoft.quantum.iqsharp.magic-ref.config> (available only in Q# Jupyter Notebooks) to tweak the format of `DumpMachine` output. It offers a lot of settings convenient in different scenarios. For example, by default `DumpMachine` uses little-endian integers to denote the basis states (the first column of the output); if you find raw bit strings easier to read, you can use `%config dump.basisStateLabelingConvention="Bitstring"` to switch.
+
+
+1. Jupyter Notebook offers the possibility to visualize the execution of the quantum program as a quantum circuit by using <xref:microsoft.quantum.iqsharp.magic-ref.trace> (available only in Q# Jupyter Notebooks). This command traces one run of the Q# program and build a circuit based on that execution. This is the circuit resulting from the running of `%trace MultiQubitDumpMachineDemo`, 
+
+![DumpMachine trace output](~/media/dumpmachine-trace-output.png)
+
+   The visualization is interactive, allowing you to click on each block to drill down to the intrinsic gates.
+
+1. Finally, <xref:microsoft.quantum.iqsharp.magic-ref.debug> (available only in Q# Jupyter Notebooks) allows you to combine tracing the program execution (as a circuit) and observing the program state as it evolves at the same time. The visualization is also interactive, you can click through each of the steps until the program run is complete, and switch to observe real and imaginary components of the amplitudes instead of measurement probabilities in the beginning of the program.
```

Comment:
```suggestion
1. Finally, <xref:microsoft.quantum.iqsharp.magic-ref.debug> (available only in Q# Jupyter Notebooks) allows you to combine tracing the program execution (as a circuit) and observing the program state as it evolves at the same time. The visualization is also interactive; you can click through each of the steps until the program run is complete, and switch to observe real and imaginary components of the amplitudes, instead of measurement probabilities in the beginning of the program.
```